src/engine/ChunkSystemTileMap.ts

```diff
@@ -0,0 +1,466 @@
+          } else {
+            removedTrailingChunks++;
+          }
+        }
+      }
+      leadingColumnsToRemove = Math.min(leadingColumnsToRemove, removedLeadingChunks);
+      trailingColumnsToRemove = Math.min(trailingColumnsToRemove, removedTrailingChunks);
+
+      if (chunkRowIndex < onScreenRowIndexStart) {
+        if (rowCleared && leadingRowsToRemove === chunkRowIndex) {
+          leadingRowsToRemove++;
+        }
+      } else {
+        if (rowCleared) {
+          trailingRowsToRemove++;
+        } else {
+          trailingRowsToRemove = 0;
+        }
+      }
+    }
+
+    this._chunks.splice(this._chunks.length - trailingRowsToRemove);
+    this._chunks.splice(0, leadingRowsToRemove);
+    this._chunksYOffset += leadingRowsToRemove;
+    for (let rowIndex = 0; rowIndex < this._chunks.length; rowIndex++) {
+      const chunkRow = this._chunks[rowIndex];
+      chunkRow.splice(chunkRow.length - trailingColumnsToRemove);
+      chunkRow.splice(0, leadingColumnsToRemove);
+    }
+    this._chunksXOffset += leadingColumnsToRemove;
+  }
+}
+
+/**
+ * The [[ChunkSystemTileMap]] class provides a way to do extremally large scenes with collision
+ * without the overhead of actors. As the name implies, the ChunkSystemTileMap is used as a regular
+ * [[TileMap]], however its cells are organized into tiled square chunks. This allows loading of the
+ * currently needed chunks on demand and unloading the currently unneeded chunks from the memory.
+ */
+export class ChunkSystemTileMap extends Configurable(ChunkSystemTileMapImpl) {}
+
+export function wrapSimpleChunkGenerator(simpleGenerator: SimpleChunkGenerator): ChunkGenerator {
+  return (chunkColumn: number, chunkRow: number, chunkSystemTileMap: ChunkSystemTileMap, engine: Engine) => {
+    const chunkCellColumn = chunkColumn * chunkSystemTileMap.chunkSize;
+    const chunkCellRow = chunkRow * chunkSystemTileMap.chunkSize;
+    const chunk = new TileMap({
+      x: chunkSystemTileMap.x + chunkCellColumn * chunkSystemTileMap.cellWidth,
+      y: chunkSystemTileMap.y + chunkCellRow * chunkSystemTileMap.cellHeight,
+      cellWidth: chunkSystemTileMap.cellWidth,
+      cellHeight: chunkSystemTileMap.cellHeight,
+      rows: chunkSystemTileMap.chunkSize,
+      cols: chunkSystemTileMap.chunkSize
+    });
+    return simpleGenerator(chunk, chunkCellColumn, chunkCellRow, chunkSystemTileMap, engine);
+  };
+}
+
+export function wrapSimpleCellGenerator(simpleGenerator: SimpleCellGenerator): ChunkGenerator {
+  return wrapSimpleChunkGenerator((chunk, chunkCellColumn, chunkCellRow, chunkSystemTileMap, engine) => {
+    const { cols, rows } = chunk;
+    for (let row = 0; row < rows; row++) {
+      for (let column = 0; column < cols; column++) {
+        simpleGenerator(chunk.getCell(column, row), chunkCellColumn + column, chunkCellRow + row, chunk, chunkSystemTileMap, engine);
+      }
+    }
+    return chunk;
+  });
+}
+
+function isSafeInteger(number: number): boolean {
```

Comment:
I have updated the Polyfill.ts and lib settings in tsconfig.json. I had to include more polyfills that originally intended since [some typescript lib definitions](https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2015.core.d.ts) are not very finely granular.